### ISSUE TITLE
etcd watcher API re-connect behavior

### ISSUE THREAD

**Issue by niyue** [etcd-cpp-apiv3/etcd-cpp-apiv3#73]
Hi there, 

For the etcd watcher API in the etcd-cpp-apiv3 library, if the connection to etcd is down for a while, will the watcher auto reconnect in this case or is it application's responsibility to re-connect? 

If it is application's responsibility to re-connect, is there any recommended way to detect such a connection failure for the watcher and re-connect? 

I cannot find any doc in the repo describing the behavior for this, could you please shed some light on it? Thanks.

**Comment by sighingnow** [comment]
> will the watcher auto reconnect in this case or is it application's responsibility to re-connect?

Its the application's responsibility for re-connectting.

> is there any recommended way to detect such a connection failure for the watcher and re-connect?

You could issue a `Wait(callback)` call after starting the watcher:

```cpp
auto watcher = ......;

watcher->Wait([this, prefix, callback](bool cancalled) {
          if (cancalled) {
            return;
          }
          // it is not cancelled
          //
          // reconnecting code here...
});
```

**Comment by sighingnow** [comment]
I will add a section in README for how to detect the failure and reconnect later.

**Comment by niyue** [comment]
Thanks. After reading your sample code, I would like to know:
1) the meaning of the arguments in the lambda, including `prefix` and `callback`. I guess `prefix` indicates the key for the watcher, and the callback indicates the callback for the watcher when constructing it. Once we find etcd is back online, we could use these information to re-construct a new watcher object, is it correct?
2) about the reconnecting code, could you give some more description about how the reconnecting code should look like? Currently, when using the `Watcher` API, there is no explicit code to `connect` but simply constructing a `Watcher`, so for `re-connecting`, how this should be done? Since the etcd server may be down for a while, I assume client should keep retrying until it finds the etcd server is back online and then create a new Watcher object, is this correct? I read the source code/doc for `etcd::Client`, there seems no direct API to verify if a connection is active, what should be the recommended approach to verify if the connection to etcd server? 

Thanks.

**Comment by sighingnow** [comment]
> Once we find etcd is back online, we could use these information to re-construct a new watcher object, is it correct?

Yes, those two refers could be used for reconnecting (renew a watcher).

> what should be the recommended approach to verify if the connection to etcd server?

Assuming you could define some global variables / class member variables

```cpp
std::string endpoints = ...;  // etcd endpoints
std::function<void(Response)> callback = ....
std::shared_ptr<etcd::Watcher> watcher;
const std::string prefix = ...;
```

And a function/member method to initialize the watcher:

```cpp
void initialize_watcher() {
    while (true) {
        # wait until the client ok again.
        Client client(endpoints);
        if (client.head().get().is_ok()) {
            break;
        }
        sleep(1);
    }
    watcher.reset(new etcd::Watcher(client, prefix, callback, true));
    watcher.Wait([endpoints, callback, &watcher, prefix](bool cancelled) {
        if (cancelled) {
            return;
        }
        initialize_watcher();
    });
}
```

To archive your goals, you could

1. watch util the endpoints is connectable by a `head()` request
2. re-initialize the watcher in the wait callback, by invoke the `initialize_watcher` again.

**Comment by niyue** [comment]
It looks like the Client's `head` API is only available after this commit (https://github.com/etcd-cpp-apiv3/etcd-cpp-apiv3/commit/efcecb7731b64a3e95d931fec061c40860ce3064), 28 days ago. However, the current release list contains only 0.2.1, which doesn't have this `head` API yet. 

I can update my local revision to the latest to take advantage of this API, but if we add such a piece of sample code to the doc, I think we should issue a new release as well so that users won't get confused where to find the `head` API.

Thanks so much for the help.

**Comment by sighingnow** [comment]
It would be great if you could submit a pull request to add the doc, as as a user with fresh mind, you could write better guide for other users than me, the developer of this repo.

I agree that we should cut a new release.

Thanks!

**Comment by niyue** [comment]
Glad to help. I would like to confirm one thing about the usage here, I assume the application code will look like (I gave it a try and it worked for me locally):
```
std::string endpoints = ...;  // etcd endpoints
std::function<void(Response)> callback = ....;
std::shared_ptr<etcd::Watcher> watcher;
const std::string prefix = ...;
initialize_watcher(endpoint, prefix, callback, watcher);
```

Here the application will be blocking for the initial connection to etcd (unless an extra thread is used in application), and the re-connection will happen in a background thread launched by the client library, is this correct?

**Comment by sighingnow** [comment]
> is this correct?

Yes.

It is also possible to avoid the blocking, as `client.head()` can be used in async style as well. But I think it doesn't matter, as the user expectation might be continuing the following job after the watcher has been established.

**Comment by niyue** [comment]
PR submitted for adding the doc. Please help to review and let me know if there is anything I need to change. Thanks.

**Comment by sighingnow** [comment]
PR merged. Closed by #76.

**Comment by niyue** [comment]
@sighingnow after more tests, I realized the code for re-connecting is still problematic.
```
void initialize_watcher(const std::string &endpoints, 
  const std::string &prefix,
  std::function<void(Response)> callback, 
  std::unique_ptr<etcd::Watcher> &watcher) {
    Client client(endpoints);
    wait_for_connection(client); 
    watcher->reset(new etcd::Watcher(client, prefix, callback));
    watcher->Wait([endpoints, prefix, callback, &watcher](bool cancelled) {
        if (cancelled) {
            return;
        }
        initialize_watcher(endpoints, prefix, callback, watcher);
    });
}
```
In this piece of code, when etcd gets restarted, the above callback will be called. When the callback is called, it will run this line of code:
```
watcher->reset(new etcd::Watcher(client, prefix, callback));
```
This will reset the old watcher pointer, leading the old watcher's destructor to be called. During destruction, the old watcher will join the `_task` thread, however, the `task_` thread is the same thread handling the `Wait` callback, which means the thread is joining itself and causes the `reset` to fail eventually.

It seems we cannot destruct the old watcher in the callback since the callback is handled by the `task_` thread which is a member variable of the old watcher. I tried moving the old watcher out of the callback to make it live longer than the `task_` thread, but it seems not easy to get right for API users. Do you have any suggestion on how this could be done for addressing this issue?

**Comment by sighingnow** [comment]
Will fix.

**Comment by niyue** [comment]
I am not sure if it is my fault to not using `shared_ptr` initially (I double checked your sample code pasted initially, and you use a `shared_ptr` for watcher).

Since I don't realize how it is implemented internally, I thought it doesn't have to be a `shared_ptr` and use `unique_ptr` instead. If I switch to use `shared_ptr` and pass it to the callback lambda for `Wait`, the `shared_ptr` won't be destructed in the callback so I think it is okay. 

Currently my code looks like this. 
```
static unique_ptr<Watcher>
  _create_watcher(const string &etcd_url, const string &key,
                  const std::function<void(const ::etcd::Response &resp)> &callback,
                  bool recursive) {
    Client client(etcd_url);
    _wait_for_connection(client);
    auto watcher = std::make_unique<Watcher>(client, key, callback, recursive);
    return watcher;
  }

initialize_watcher(const string &etcd_url, const string &key,
                     const std::function<void(const ::etcd::Response &resp)> &value_change_callback,
                     bool recursive, shared_ptr<Watcher> &watcher ) {
    auto new_watcher = _create_watcher(etcd_url, key, value_change_callback, recursive);
    new_watcher->Wait([etcd_url, key, value_change_callback, recursive, watcher](bool cancelled) mutable {
      _logger->info("etcd watcher disconnected from etcd", "etcd_url", etcd_url, "key", key);
      if (!cancelled) {
        initialize_watcher(etcd_url, key, value_change_callback, recursive, watcher);
      }
    });
    watcher = std::move(new_watcher);
  }
```

Please let me know if there is better alternative than this approach. Thanks.

**Comment by sighingnow** [comment]
> I am not sure if it is my fault to not using `shared_ptr` initially (I double checked your sample code pasted initially, and you did use a `shared_ptr` for watcher).
> 
> Since I don't realize how it is implemented internally, I thought it doesn't have to be a `shared_ptr` and use `unique_ptr` instead. If I switch to use `shared_ptr` and pass it to the callback lambda for `Wait`, the `shared_ptr` won't be destructed in the callback so I think it is okay.
> 
> Currently my code looks like this.
> 
> ```
> static unique_ptr<Watcher>
>   _create_watcher(const string &etcd_url, const string &key,
>                   const std::function<void(const ::etcd::Response &resp)> &callback,
>                   bool recursive) {
>     Client client(etcd_url);
>     _wait_for_connection(client);
>     auto watcher = std::make_unique<Watcher>(client, key, callback, recursive);
>     return watcher;
>   }
> 
> initialize_watcher(const string &etcd_url, const string &key,
>                      const std::function<void(const ::etcd::Response &resp)> &value_change_callback,
>                      bool recursive, shared_ptr<Watcher> &watcher ) {
>     auto new_watcher = _create_watcher(etcd_url, key, value_change_callback, recursive);
>     new_watcher->Wait([etcd_url, key, value_change_callback, recursive, watcher](bool cancelled) mutable {
>       _logger->info("etcd watcher disconnected from etcd", "etcd_url", etcd_url, "key", key);
>       if (!cancelled) {
>         initialize_watcher(etcd_url, key, value_change_callback, recursive, watcher);
>       }
>     });
>     watcher = std::move(new_watcher);
>   }
> ```
> 
> Please let me know if there is better alternative than this approach. Thanks.

I don't think it fixes the problem (if I understand it correctly), as it still needs to destruct itself in the callback. At the end of the wait callback (the end of `initialize_watcher`) you still need to destruct the previous watcher, however it happens in the thread luanched by previous watcher.

**Comment by sighingnow** [comment]
I think we should enable "re-connect" in the wait callback, rather than creating a new watcher. Does that make sense?

**Comment by niyue** [comment]
Yes, I think it will make the library user's life much easier. Simply a toggle indicating if the watcher requiring re-connect or not is much easier to understand.

**Comment by sighingnow** [comment]
> Actually in my code, I wrap initialize_watcher into a class so that the entire app using this watcher API will have the capability to re-connect.

But the previous watcher is still destructed in `watcher = std::move(new_watcher);`, which will `join` the current thread, right?

**Comment by niyue** [comment]
In my current code, the `watcher` is a `shared_ptr` and `new_watcher` is created as a `unique_ptr` and stored externally as a `shared_ptr`, since the previous `watcher` is passed into the `Wait` lambda, `watcher = std::move(new_watcher)` doesn't trigger the destruction of `watcher` (at least to my current testing)

**Comment by sighingnow** [comment]
> since the previous watcher is passed into the Wait lambda, watcher = std::move(new_watcher) doesn't trigger the destruction of watcher (at least to my current testing)

I see. thanks for the clarification.

**Comment by sighingnow** [comment]
I have updated the doc in 891dc7bedf926e200cda171d7f4c0ca59f276736 for the `std::shared_ptr` usage.

The "reconnect" in the watcher itself part will be fixed later.

**Comment by sighingnow** [comment]
The previous problematical example has been fixed in #123 and the README has been revisited.
